### DISCUSSION TITLE
Is it possible to get the x, y location of the caret in a TextField?

### DISCUSSION THREAD

**Discussion by manishpatel0404** [airsdk/Adobe-Runtime-Support#1985] [Q&A] [answered]
If it is possible to get the x, y location of the caret in a `TextField`, it would make it possible to create custom scollbar schemes for a TextField. The built-in scrollbar does not seem very appealing in custom themes.

Not very earth-shattering, but it would be nice to have this.
Thanks.

**Comment by MalacTheLittle** [comment, marked as answer]
It is possible, you should check it in AS3 [`TextField`](https://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/flash/text/TextField.html) documentation. You can use `getCharBoundaries` method and as _charIndex_ use `caretIndex` property.

Here's an example:
```
input.addEventListener(Event.ENTER_FRAME, track);
function track(event:Event):void{
	var caretIndex = input.caretIndex;
	var rect = input.getCharBoundaries(caretIndex);
	  
	trace("cIndex:" + caretIndex + " x:" + (rect?rect.x:"null") + " y:" + (rect?rect.y:"null"));
}
```

**Reply by manishpatel0404** [reply]
It should be `getCharBoundaries(caretIndex - 1)` when the caretIndex is after the last character in the text, but otherwise, ok.

The only problem is when there are blank lines in the text, in which case, `getCharBoundaries` returns a `null` rectangle.

Thanks.

**Reply by MalacTheLittle** [reply]
Yes, also the problem is with empty lines where `caretIndex - 1` doesn't help. You should consider using the `getLineMetrics` method in that case:

```
if(!rect){
	rect = {x:0, y:0};
	var caretLine = input.getLineIndexOfChar(caretIndex);
	if(caretLine < 0){caretLine = (caretIndex == input.text.length ? input.numLines-1 : 0);}
	for(var i:uint = 0; i < caretLine; ++i){
		rect.y += input.getLineMetrics(i).height;
	}
}
```

**Reply by manishpatel0404** [reply]
Finally, got the custom scrollbar working with `TextField` using `getLineMetrics`, `scrollV` and `scrollH`.

Thanks, @MalacTheLittle for pointing me to `getLineMetrics`. I realised that I did not have to use `caretIndex` at all.
